### PR TITLE
[Mongo]: Add MongoDB

### PR DESCRIPTION
**Pull request by jondeaves**
## Context

Allowing the bot to store its own data will help make better commands.

## Description

- Adds a MongoDB connection to the bot. Connects on start and provides full CRUD.
- Each database command must also receive the user id that triggers the command, as a type of audit log.
- Adds new command that allows user to add messages to a 'greetings' collection. But does not have the server greeting use this data yet.

## Examples

```
findOne('123456', 'collectionName', { ident: 'generated-slug' })
find('123456', 'collectionName', { key: 'value' })
insert('123456', 'collectionName', [{ body: 'This is a document!' }])
updateMany('123456', 'collectionName', { ident: 'generated-slug' }, { secondKey: 'new data'})
deleteMany('123456', 'collectionName', { ident: 'generated-slug' })
```

## Testing

- [x] Manual

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -44,7 +53,7 @@ export default class App {
       };
 
       try {
-        await command.execute(message, args, prefix, commandList);
+        await command.execute(message, args, prefix, commandList, this._dbService);
```
Comment (chenasraf) [inline src/app.ts:56]: Not necessarily right now, but maybe we should consider passing some of the arguments into an object? It seems they might continue piling up, and we might not always want to take a lot of unnecessary args in our commands

File: src/core/services/mongo.service.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+   * Fetches the first document that matches the query
+   *
+   * @param collection String identifier for collection document belongs to
+   * @param query key/value pairs of search conditions
+   *
+   * @returns Returns matched document
+   *
+   * @example findOne('123456', 'collectionName', { ident: 'generated-slug' })
+   */
+  public async findOne(userId: string, collection: string, query: any) {
```
Comment (chenasraf) [inline src/core/services/mongo.service.ts:54]: Where do we use userId? Why is it needed here? I believe user permissions should be on a higher level, somewhere, if that was the purpose of this.

File: src/bot/commands/addgreeting.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  example: `\`${prefix}addgreeting Welcome to the club {name}\``,
+  async execute(message: Discord.Message, args: string[], prefix?: string, commands?: Collection<string, ICommand>, dbService?: MongoService) {
+    let isPermitted = false;
+    const permittedRoles = ['bot-devs', 'admin', 'staff']
+
+    message.guild.roles.cache.forEach(r => {
+      if (permittedRoles.indexOf(r.name) !== -1) {
+        isPermitted = true;
+      }
+    })
```
Comment (chenasraf) [inline src/bot/commands/addgreeting.ts]: Two points:
1. This can be shortened and short-circuited, by using `Array.prototype.some`
1. I believe in Discord.js ([docs](https://discord.js.org/#/docs/main/stable/class/Message?scrollTo=guild)) the `message.guild.roles` object doesn't give you the guilds of the user, but the roles of the guild. If you wanna check permissions, we should use `message.member.roles.cache` [[cit1]](https://stackoverflow.com/questions/56926998/how-can-i-check-if-the-message-author-has-an-admin-role-using-discord-js) [[cit2]](https://discord.js.org/#/docs/main/stable/class/GuildMember?scrollTo=roles)

In summary, this should be what we are looking for:
```suggestion
    const isPermitted = message.member.roles.cache.some(r => permittedRoles.indexOf(r.name) !== -1);
```

This also prevents the unnecessary runs over the permissions after the flag is found.
Do we want to move this behavior into a utility? We will probably be using it a lot.

Check my suggestion code works, I haven't tested it. But it should.

**Review by chenasraf — CHANGES_REQUESTED** [review]
Looks great in general, I have a couple of comments, check them out :)

File: src/app.ts (hunk 1 above)
Comment (jondeaves) [reply src/app.ts:56]: Yeah that was my thinking too, it's a simple approach right now that won't scale so was thinking of enhancing the command creation process when possible.

File: src/core/services/mongo.service.ts (hunk 2 above)
Comment (jondeaves) [reply src/core/services/mongo.service.ts:54]: It's for the logging library, so that there is a userId attached to each database command that happens as well. Down the line I'd prefer to separate it into an audit system, separate from the logger, probably attached to commands.

**Review by chenasraf — APPROVED** [review]
